File: leetcode_problems/cpp/lc104_max_depth_of_binary_tree.cpp
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int depth(TreeNode *node, int d) {
        if (node == nullptr) {
            return d;
        } else {
            int left = depth(node->left, d+1);
            int right = depth(node->right, d+1);
            if (left > right) {
                return left;
            } else {
                return right;
            }
        }
    }

    int maxDepth(TreeNode* root) {
        return depth(root, 0);
    }
};

File: leetcode_problems/typescript/src/lc104_max_depth_of_binary_tree.ts
function maxDepthRecursion(node: TreeNode | null, d: number): number {
    if (node === null) {
        return d;
    }
    let left = maxDepthRecursion(node.left, d + 1) ;
    let right = maxDepthRecursion(node.right, d+1);
    return left > right ? left : right;
}

function maxDepth(root: TreeNode | null): number {
    return maxDepthRecursion(root, 0);
};